# HG changeset patch
# User Mike Miller <mtmiller@octave.org>
# Date 1506283291 25200
#      Sun Sep 24 13:01:31 2017 -0700
# Node ID 8f3db1ac787747552083d35ebdc3cf065a10a379
# Parent  eadaf557ca099ecc83d4187d0c2fe249278f9f21
tar, untar, unpack: ensure TAR_OPTIONS doesn't affect behavior (bug #52095)

* tar.m: Add unwind_protect block to save and restore the TAR_OPTIONS
environment variable.
* unpack.m: Save and restore the TAR_OPTIONS environment variable within the
existing unwind_protect block.

diff --git a/scripts/miscellaneous/tar.m b/scripts/miscellaneous/tar.m
--- a/scripts/miscellaneous/tar.m
+++ b/scripts/miscellaneous/tar.m
@@ -58,17 +58,26 @@ function filelist = tar (tarfile, files,
   if (ispc)
     ## Change tarfile into a mingw style acceptable for tar
     tarfile = __w2mpth__ (tarfile);
   endif
 
   cmd = sprintf ("tar cvf %s -C %s %s",
                           tarfile, rootdir, sprintf (" %s", files{:}));
 
-  [status, output] = system (cmd);
+  ## Save and restore the TAR_OPTIONS environment variable used by GNU tar.
+  tar_options_env = getenv ("TAR_OPTIONS");
+  unwind_protect
+    unsetenv ("TAR_OPTIONS");
+    [status, output] = system (cmd);
+  unwind_protect_cleanup
+    if (! isempty (tar_options_env))
+      setenv ("TAR_OPTIONS", tar_options_env);
+    endif
+  end_unwind_protect
 
   if (status)
     error ("tar: tar exited with status = %d", status);
   endif
 
   if (nargout > 0)
     filelist = ostrsplit (output, "\r\n", true);
     filelist = filelist';
diff --git a/scripts/miscellaneous/unpack.m b/scripts/miscellaneous/unpack.m
--- a/scripts/miscellaneous/unpack.m
+++ b/scripts/miscellaneous/unpack.m
@@ -235,21 +235,27 @@ function filelist = unpack (file, dir = 
     [status, msg] = mkdir (dir);
     if (! status)
       error ("unpack: mkdir failed to create %s: %s", dir, msg);
     endif
   elseif (! S_ISDIR (s.mode))
     error ("unpack: %s: not a directory", dir);
   endif
 
+  ## Save and restore the TAR_OPTIONS environment variable used by GNU tar.
+  tar_options_env = getenv ("TAR_OPTIONS");
   unwind_protect
+    unsetenv ("TAR_OPTIONS");
     cd (dir);
     [status, output] = system (sprintf ([command " 2>&1"], file));
   unwind_protect_cleanup
     cd (origdir);
+    if (! isempty (tar_options_env))
+      setenv ("TAR_OPTIONS", tar_options_env);
+    endif
   end_unwind_protect
 
   if (status)
     error ("unpack: unarchiving program exited with status: %d\n%s",
            status, output);
   endif
 
   if (nargout > 0 || needmove)
